README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# rppo

[![Build
Status](https://travis-ci.org/ropensci/rppo.svg?branch=master)](https://travis-ci.org/ropensci/rppo)
[![codecov.io](https://codecov.io/github/r-lib/covr/coverage.svg?branch=master)](https://codecov.io/github/r-lib/covr?branch=master)
[![](https://badges.ropensci.org/207_status.svg)](https://github.com/ropensci/software-review/issues/207)
[![R-CMD-check](https://github.com/ropensci/rppo/workflows/R-CMD-check/badge.svg)](https://github.com/ropensci/rppo/actions)

The global plant phenology data portal, is an aggregation of plant
phenological observations from
[USA-NPN](https://www.usanpn.org/usa-national-phenology-network),
[NEON](https://www.neonscience.org/), and
[PEP725](https://www.pep725.eu/) representing 20 million phenological
observations from across North America and Europe. The PPO data portal
utilizes the [Plant Phenology
Ontology](https://github.com/PlantPhenoOntology/ppo/) (PPO) to align
phenological terms and measurements from the various databases. The rppo
R package enables programmatic access to all data contained in the PPO
data portal incuding selected classes contained in the PPO itself.

For information on how data is assembled for the PPO data portal, visit
the [ppo-data-pipeline git
repository](https://github.com/biocodellc/ppo-data-pipeline).

## Installation

You can install rppo from github with:

``` r
install.packages("devtools")
devtools::install_github("ropensci/rppo")
library(rppo)
```

## Examples

Following are a couple of brief examples to illustrate how to get
started with rppo. We recommend visiting the [rppo
vignette](https://htmlpreview.github.io/?https://github.com/ropensci/rppo/blob/master/vignettes/rppo-vignette.html)
for a more complete set of examples on using the rppo package, as well
as viewing man pages for rppo functions in the R environment, using
`?ppo_data` and
`?ppo_terms`.

``` r
# query all results from day 1 through 100 in a particular bounding box, 
# limited to 2 records
r <- ppo_data(fromDay = 1, toDay = 100, bbox="37,-120,38,-119", limit=2, timeLimit=10)
#> sending request for data ...
#> https://www.plantphenology.org/api/v2/download/?q=%2Blatitude:>=37+AND+%2Blatitude:<=38+AND+%2Blongitude:>=-120+AND+%2Blongitude:<=-119+AND+%2BdayOfYear:>=1+AND+%2BdayOfYear:<=100+AND+source:USA-NPN,NEON&source=latitude,longitude,year,dayOfYear,termID&limit=2

# view the data returned
print(r$data)
#>   dayOfYear year   genus specificEpithet latitude longitude
#> 1        33 2017 Quercus       douglasii 37.11144 -119.7315
#> 2        96 2017  Bromus        diandrus 37.11144 -119.7315
#>                                                                                                                                                                                                                                            termID
#> 1                                                                                                                                                                                                 obo:PPO_0002610,obo:PPO_0002013,obo:PPO_0002000
#> 2 obo:PPO_0002601,obo:PPO_0002610,obo:PPO_0002005,obo:PPO_0002604,obo:PPO_0002605,obo:PPO_0002013,obo:PPO_0002003,obo:PPO_0002000,obo:PPO_0002602,obo:PPO_0002006,obo:PPO_0002007,obo:PPO_0002004,obo:PPO_0002008,obo:PPO_0002603,obo:PPO_0002600
#>   source
#> 1   NEON
#> 2   NEON
#>                                                              eventId
#> 1 https://n2t.net/ark:/21547/Amn2cd982ca2-6147-4a63-a864-f4e556420562
#> 2 https://n2t.net/ark:/21547/Amn2d1a3e6de-7885-404f-828f-9ebf63248d68

# view the number of possible records returned
print(r$number_possible)
#> [1] 7251

# return a data frame of present
presentTerms <- ppo_terms(present = TRUE, timeLimit=10)
#> sending request for terms ...

# print the 2nd present term returned
print(presentTerms[2,])
#>            termID                            label
#> 2 obo:PPO_0002358 abscised fruits or seeds present
#>                                                                                                                                                                                                                                                                                                                                  definition
#> 2 An 'abscised fruit or seed presence' (PPO:0002059) trait that is a 'quality of' (RO:0000080) a 'whole plant' (PO:0000003) from which at least one 'ripe fruit' (PPO:0001045) has been abscised or removed by an herbivore or that has at least one 'ripe fruit' (PPO:0001045) that has abscised at least one 'mature seed' (PPO:0001024).
#>                                          uri
#> 2 https://purl.obolibrary.org/obo/PPO_0002358
```

## Citation

To cite the ‘rppo’ R package in publications
use:

``` 
   'John Deck, Brian Stucky, Ramona Walls, Kjell Bolmgren, Ellen Denny, Salix Dubois, Robert Guralnick' (2022). rppo: An interface to the Plant Phenology Ontology and associated data store.  R package version 2.0
   https://github.com/ropensci/rppo
```

## Code of Conduct

View our [code of conduct](https://github.com/ropensci/rppo/blob/master/CONDUCT.md)

[![ropensci\_footer](https://ropensci.org/public_images/github_footer.png)](https://ropensci.org)


NEWS.md
# rppo 2.0
## New Features
 * updated server end-point. implemented a host of code fixes.



R/ppo_data.R
#' @title Access Data From the Global Plant Phenology Data Portal
#'
#' @description Access data from the global plant phenology data portal
#' (PPO data portal)
#'
#' @param scientificName (character) A plant species scientific name.
#' @param genus (character) A plant genus name. See details.
#' @param specificEpithet (character) A plant specific epithet
#' @param termID (character) A single termID from the plant phenology ontology.
#' See details.
#' @param fromYear (integer) return data from the specified year
#' @param toYear (integer) return data up to and including the specified year
#' @param fromDay (integer) return data starting from the specified day
#' @param toDay (integer) return data up to and including the specified day
#' @param bbox (character) return data within a bounding box. Format is
#' \code{lat,long,lat,long} and is structured as a string.  Use this website:
#' http://boundingbox.klokantech.com/ to quickly grab a bbox (set format on
#' bottom left to csv and be sure to switch the order from
#' long, lat, long, lat to lat, long, lat, long).
#' @param subSource (character) return data from the specified sub-source.
#' See details.
#' @param status (character) Either "present" or "absent". Return data with the
#' specified status.
#' @param mapped_traits (character) return data from the specified traits. See
#' details
#' @param eventRemarks (character) return data from the specified eventRemarks
#' @param limit (integer) limit returned data to a specified number of records
#' @param timeLimit (integer) set the limit of the amount of time to wait for a response
#' @param keepData (logical) whether to keep (TRUE) or delete (FALSE; default)
#' the downloaded data (~/ppo_download/).
#'
#' @details
#' The ppo_data function returns a list containing the following information:
#' a readme file, citation information, a data frame with data, an integer with
#' the number of records returned and a status code. The function is called with
#' parameters that correspond to values contained in the data itself which act
#' as a filter on the returned record set. For a list of available mapped_traits,
#' termID, subSource and subSource see the \code{\link{ppo_filters}} dataset. For mapped_traits and
#' termID, the \code{\link{ppo_get_terms}} function will return a data.frame with present,
#' absent or both terms and traits information. The \code{\link{ppo_terms}} will
#' do the same but will use the API to get the lastest data. However, some of
#' the traits/termID may not return any results from this function.
#' See their documentation for more details.
#' @export
#' @keywords data download plant phenology
#' @importFrom utils read.csv
#' @importFrom utils unzip
#' @importFrom httr GET
#' @importFrom httr content
#' @importFrom readr read_file

#' @return A list with the following elements:
#' \itemize{
#'  \item {`data`: A data frame containing data}
#'  \item {`readme`: A string with information about the return package}
#'  \item {`citation`: A string with citation information}
#'  \item {`number_possible`: An integer with total possible results}
#'  \item {`status_code`: An integer with status code returned from server}
#'}
#'
#' @examples
#'
#' r1 <- ppo_data(genus = c("Quercus", "Pinus"), termID='obo:PPO_0002313',
#' limit=10, timeLimit = 4)
#' head(r1$data)
#'
#' r2 <- ppo_data(fromDay = 1, toDay = 100, bbox="37,-120,38,-119", limit=10,
#' timeLimit = 4)
#' head(r2$data)

ppo_data <- function(
    scientificName = NULL,
    genus = NULL,
    specificEpithet = NULL,
    termID = NULL,
    fromYear = NULL,
    toYear = NULL,
    fromDay = NULL,
    toDay = NULL,
    bbox = NULL,
    subSource = NULL,
    status = NULL,
    mapped_traits = NULL,
    eventRemarks = NULL,
    limit = 100000L,
    timeLimit = 60,
    keepData = FALSE) {

  # check arguments
  limit <- assert(limit, c("integer", "numeric", "character"))
  if (!length(limit) || as.integer(limit) > 100000) {
    limit <- 100000L
    warning("Maximum value for limit is 100000. Limiting query at 100000.")
  }
  assert(timeLimit, c("integer", "numeric", "character"))
  assert(keepData, "logical")
  # Check for minimum arguments to run a query
  params <- bc(list(
    scientificName = assert(scientificName, "character"),
    genus = assert(genus, "character"),
    specificEpithet = assert(specificEpithet, "character"),
    termID = assert(termID, "character"),
    fromYear = assert(fromYear, c("integer", "numeric", "character")),
    toYear = assert(toYear, c("integer", "numeric", "character")),
    fromDay = assert(fromDay, c("integer", "numeric", "character")),
    toDay = assert(toDay, c("integer", "numeric", "character")),
    bbox = assert(bbox, "character"),
    subSource = assert(subSource, "character"),
    status = assert(status, "character"),
    eventRemarks = assert(eventRemarks, "character"),
    mapped_traits = assert(mapped_traits, "character")
  ))
  if (!length(params)) {
    stop("Please specify at least 1 query argument")
  }
  if (length(status) && !status %in% c("present", "absent")) {
    stop("If specified, `status` must be one of 'present' or 'absent'.")
  }
  # for some reason multiple epiphet doesn't work with multiple genus but does
  # with multipule scientificNames
  if (length(genus) > 1 && length(specificEpithet) > 1) {
    if (!length(scientificName)) {
      params$scientificName <- params$genus
      params$genus  <- NULL
    } else {
      warning("Results might not include all queried genus.
              Try specifying genus in scientificName if that's the case.")
    }
  }
  params$source <- "USA-NPN"

  queryURL <- make_queryURL(params = params, limit = limit)

  results <- get_url(queryURL, timeLimit)

  process_response(results = results, keepData = keepData)
}

make_queryURL <- function(params, limit = 100000L) {
  # construct the value following the "q" key
  q <- "q="
  # counter to tell us if we're after 1st record
  counter <- 0
  # loop through all user parameters
  for (key in names(params)) {
    value <- params[[key]]
    # For multiple arguments, insert AND separator.  Here, we insert html
    if (counter > 0) {
      q <- paste(q, "  AND  ", sep = "")
    }
    # Begin arguments using  key:value and html encode the   sign with
    # Here, we insert html encoding   for spaces
    if (key == "fromYear")
      q <- paste0(q,' year:>=', value)
    else if (key == "toYear")
      q <- paste0(q,' year:<=', value)
    else if (key == "fromDay")
      q <- paste0(q,' dayOfYear:<=', value)
    else if (key == "toDay")
      q <- paste0(q,' dayOfYear:<=', value)
    else if (key == "termID")
      q <- paste(
        q,' ','termID',':"',value,'"', sep = "")
    else if (key == "bbox") {
      coord <- as.numeric(strsplit(value, ",")[[1]])
      lats <- sort(coord[1:3])
      lngs <- sort(coord[2:4])
      q <- paste0(q,
                  " latitude:>=", lats[1],
                  " AND  latitude:<=", lats[2],
                  " AND  longitude:>=", lngs[1],
                  " AND  longitude:<=", lngs[2]
      )
    } else {
      if (length(value) > 1) {
        if (key == "scientificName" || key == "specificEpithet")
          value <- paste(value, collapse = "|")
        else
          value <- paste0("(", paste0('"', value, '"', collapse = ","),")")
      }
      q <- paste0(q, ' ', key, ':', value)
    }
    counter <- counter + 1
  }

  queryURL <- paste0("https://biscicol.org/api/v3/download/_search?", q,
                     '&limit=', limit)
  gsub('"', '\\"', gsub(" ", "+", queryURL))
}

process_response <- function(results, keepData = FALSE) {
  if (!length(results) && results$status_code != 200)
      list(
        "data" = NULL,
        "readme" = NULL,
        "citation" = NULL,
        "number_possible" = 0,
        "status_code" = results$status_code)
  else {
    tf <- tempfile()

    # save file to disk
    writeBin(httr::content(results, "raw"), tf)
    unzip(tf, exdir = "~/ppo_download")

    # data.csv contains all data as comma separated values
    ppo <- utils::read.csv('~/ppo_download/data.csv', header = TRUE)

    # README.txt contains information about the query and the # of results
    readme <- readr::read_file('~/ppo_download/README.txt')
    class(readme) <- append(class(readme), "ppo_text")

    # citation_and_data_use_policies.txt contains citation information
    citation <- readr::read_file('~/ppo_download/citation_and_data_use_policies.txt')
    class(citation) <- append(class(citation), "ppo_text")

    # grab the number of possible from the readme file
    numPossible <- gsub("^.*possible = |\ntotal results returned.*$|,",
                        "\\1", readme)
    unlink(tf)
    if (!keepData) unlink("ppo_download/", recursive = TRUE)

    list(
      "data" = ppo,
      "readme" = readme,
      "citation" = citation,
      "number_possible" = as.integer(numPossible),
      "status_code" = results$status_code)
  }
}
